compiler/damlc/daml-stdlib-src/DA/Internal/Template/Functions.daml

```diff
@@ -14,75 +14,115 @@ daml 1.2
   agreement : t -> Text
 
+class HasCreate t where
   -- | Create a contract based on a template `t`.
   create : t -> Update (ContractId t)
 
+class HasFetch t where
   -- | Fetch the contract data associated with the given contract ID.
   -- If the `ContractId t` supplied is not the contract ID of an active
   -- contract, this fails and aborts the entire transaction.
   fetch : ContractId t -> Update t
 
+class HasArchive t where
   -- | Archive the contract with the given contract ID.
   archive : ContractId t -> Update ()
 
-  -- | Wrap the template in AnyTemplate
-  toAnyTemplate : t -> AnyTemplate
-
-  -- | Extract the underlying template from AnyTemplate if the type matches
-  -- or return `None`.
-  fromAnyTemplate : AnyTemplate -> Optional t
-
-  -- | Generate a unique textual representation of the template Id.
-  _templateTypeRep : proxy t -> TemplateTypeRep
-
 -- | The stakeholders of a contract: its signatories and observers.
-stakeholder : Template t => t -> [Party]
+stakeholder : (HasSignatory t, HasObserver t) => t -> [Party]
 stakeholder t = signatory t ++ observer t
 
-class Template t => Choice t c r | t c -> r where
+-- | Constraints satisfied by template choice types.
+type Choice t c r =
+  ( Template t
+  , HasExercise t c r
+#ifdef DAML_TYPE_REP && DAML_ANY_TYPE
+  , HasToAnyChoice t c r
+  , HasFromAnyChoice t c r
+#endif
+  )
+
+class HasExercise t c r | t c -> r where
   -- | Exercise a choice on the contract with the given contract ID.
   exercise : ContractId t -> c -> Update r
-  _toAnyChoice : proxy t -> c -> Any
-  _fromAnyChoice : proxy t -> Any -> Optional c
 
-class Template t => TemplateKey t k | t -> k where
+-- | Constrants satisfied by template key types.
+type TemplateKey t k =
+  ( Template t
+  , HasKey t k
+  , HasLookupByKey t k
+  , HasFetchByKey t k
+  , HasMaintainer t k
+#ifdef DAML_TYPE_REP && DAML_ANY_TYPE
+  , HasToAnyContractKey t k
+  , HasFromAnyContractKey t k
+#endif
+  )
+
+class HasKey t k | t -> k where
   -- | The key of a contract.
   key : t -> k
 
+class HasLookupByKey t k | t -> k where
   -- | Look up the contract ID `t` associated with a given contract key `k`.
   --
   -- You must pass the `t` using an explicit type application. For
   -- instance, if you want to look up a contract of template `Account` by its
   -- key `k`, you must call `lookupByKey @Account k`.
   lookupByKey : k -> Update (Optional (ContractId t))
 
+class HasFetchByKey t k | t -> k where
   -- | Fetch the contract ID and contract data associated with a given
   -- contract key.
   --
   -- You must pass the `t` using an explicit type application. For
   -- instance, if you want to fetch a contract of template `Account` by its
   -- key `k`, you must call `fetchByKey @Account k`.
   fetchByKey : k -> Update (ContractId t, t)
-  -- NOTE(FM): the motivation for this function to return both the
+
+class HasMaintainer t k | t -> k where
+  -- NOTE(F. Mazzoli): the motivation for this function to return both the
```

Comment:
It looks like this comment belongs to `fetchByKey`.

Reply:
Oops, fixed!